casper/src/main/scala/coop/rchain/casper/engine/Running.scala

```diff
@@ -201,20 +202,27 @@ object Running {
         )
     )
 
-  def handleBlockMessage[F[_]: Monad: Log: RequestedBlocks](peer: PeerNode, b: BlockMessage)(
+  def handleBlockMessage[F[_]: Monad: Log: RequestedBlocks](
+      peer: PeerNode,
+      b: BlockMessage
+  )(
       repeatedCasperMessage: BlockHash => F[Boolean],
+      isInBlockStore: BlockHash => F[Boolean],
       casperAdd: BlockMessage => F[ValidBlockProcessing]
-  ): F[Unit] =
-    repeatedCasperMessage(b.blockHash)
+  ): F[Unit] = {
+    val alreadyInStore = isInBlockStore(b.blockHash)
+    (repeatedCasperMessage(b.blockHash) ||^ alreadyInStore)
       .ifM(
-        Log[F].info(s"Received block ${PrettyPrinter.buildString(b.blockHash)} again."),
+        Log[F].info(s"Received block ${PrettyPrinter.buildString(b.blockHash)} again.") >>
+          alreadyInStore.ifM(RequestedBlocks.remove(b.blockHash), ().pure[F]),
```

Comment:
@nzpr I seen messages re-requested even when in the block store so that's why they are removed here also.

Reply:
Logic was that hash should be be removed from RequestedBlocks only after it is added to DAG. But we should revisit this logic

Reply:
This should not be done, because if block in store it might be not in process of adding to casper - nor added to casper